src/absform.erl

```diff
@@ -27,13 +29,15 @@ normalize_record_field({typed_record_field,
 
 %% @doc Turns all function types into bounded function types. Add default empty
 %% constraints if missing.
+-spec normalize_function_type_list(FunTypeList) -> FunTypeList when
+      FunTypeList :: gradualizer_type:af_function_type_list().
 normalize_function_type_list(FunTypeList) ->
-    lists:map(fun normalize_function_type/1, FunTypeList).
+    ?assert_type(lists:map(fun normalize_function_type/1, FunTypeList), [B, ...]).
```

Comment:
It'd be nice to propagate the fact that the list is non-empty through functions like `lists:map`. It would likely need special handling in the type checker though, similar to how we treat arithmetic operators.

Reply:
I think we could rely on multiple spec clauses to do it like this:

```erlang
-spec lists:map(fun(), nonempty_list()) -> nonempty_list();
               (fun(), list()) -> list().
```

Then it would be a matter of providing a few overrides in prelude. The same would be nice for proper and improper lists - we touched on that in the thread of https://github.com/josefs/Gradualizer/pull/440#discussion_r957586771.

I think we just have to make sure the function intersection types work well enough to handle such specs. I have something in progress in #461. In that draft this code

```erlang
-spec foo(bar) -> bar;
         (list()) -> baz | qux.
foo(bar) -> bar;
foo([]) -> baz;
foo([a]) -> qux.
```

is already accepted, but this

```erlang
-type t() :: {tag, any()}.

-spec f(t()) -> string();
       ([t()]) -> string().
f(Types) when is_list(Types) -> lists:map(fun f/1, Types);
f({tag, _}) -> "".
```

still fails. Still working on it.

Reply:
`lists:map(fun f/1, Types)` returns a nested list of strings, which isn't in line with the spec. Is this why it fails?

Reply:
No, no, it fails checking the intersection:

```
test/should_pass/spec_and_fun_clause_intersection_pass.erl:
The pattern {tag, _} on line 16 at column 3 doesn't have the type [t()]
```

Reply:
Good point anyway! I think we'll only catch that with the constraint solver in place, though.